### ISSUE TITLE
When setting a new PK in setup mode, PK should not be required to be self-signed (Bugzilla Bug 2506)

### ISSUE THREAD

**Issue by tianocore-issues** [tianocore/edk2#8350]
This issue was created automatically with [bugzilla2github](https://github.com/makubacki/bugzilla2github/tree/bz2gh_tianocore)
# Bugzilla Bug 2506
Date: 2020-02-06T21:26:02+00:00
From: Matthew Carlson &lt;<macarl>&gt;
To:   Jan Bobek &lt;<jbobek>&gt;
CC:   @corthon, @lgao4, jbobek, @jyao1

Last updated: 2023-02-04T06:54:22+00:00

**Comment by tianocore-issues** [comment]
## Comment 11148
Date: 2020-02-06 21:26:02 +0000
From: Matthew Carlson &lt;<macarl>&gt;

- Industry Specification: ---
- Release Observed: EDK II Master
- Releases to Fix: EDK II Master
- Target OS: ---
- Bugzilla Assignee(s): Jan Bobek &lt;<jbobek>&gt;

In ProcessVarWithPk in AuthService.c, we check fall into the last else branch if we are in SETUP mode and it is the PK that we&apos;re updating. The PK packet shouldn&apos;t be self-signed.

**Comment by tianocore-issues** [comment]
## Comment 11164
Date: 2020-02-07 07:08:28 +0000
From: @lersek

Matthew,

The following mailing list thread (a patch posting from 2019) is related:

* [edk2-devel] [PATCH]
  SecurityPkg: Don&apos;t Verify the enrolled PK in setup mode

  http://mid.mail-archive.com/7564.1562045108414671150@groups.io
  https://edk2.groups.io/g/devel/message/43150

The patch there removes the last &quot;else&quot; branch that you mention.

The patch was not been merged; I objected to it in its posted form. A v2 was requested (with a FeaturePCD), but then Derek Lin (the original submitter) decided he liked the in-tree behavior more than a new FeaturePCD.

To summarize my request from back then: I don&apos;t object to the proposed PK enrollment relaxation as an option (even as the default option), I&apos;d just like a FeaturePCD for platforms that prefer keeping the current self-signedness requirement. Thanks!

**Comment by tianocore-issues** [comment]
## Comment 11168
Date: 2020-02-07 13:34:25 +0000
From: Matthew Carlson &lt;<macarl>&gt;

Laszlo,

Thanks for commented with those links, it&apos;s very helpful to see more of the history around this issue. I agree- a FeaturePCD is needed here. Should the disabled behavior to be the same as what it currently does now? Since we the spec says that it doesn&apos;t have to be checked for self-signing, should the PCD be opt-in or opt-out?

**Comment by tianocore-issues** [comment]
## Comment 11181
Date: 2020-02-10 07:28:00 +0000
From: @lersek

Hi Matthew,

I think a platform should not be required to tweak a PCD (i.e., to diverge from the PCD default inherited from the package DEC file), in order to get spec-conformant behavior.

In other words, the default behavior for the logic in question should be changed (relaxed) so that it match the spec. And then the PCD could be used by platforms to diverge from the spec (i.e., in this particular case, to enforce self-signedness on the PK, which is stricter than what the spec mandates).

So whoever proposed a patch for SecurityPkg, I&apos;d ask them to attach two more patches (one for OvmfPkg and ArmVirtPkg each), for opting into the strict self-signedness check.

Whether the name of the new PCD should be in the negative sense (&quot;disable self-signedness check&quot;) or in the positive sense (&quot;enable self-signedness check&quot;), is an orthogonal question. I&apos;ll defer to the SecurityPkg maintainers on that.

Thanks!
Laszlo

**Comment by tianocore-issues** [comment]
## Comment 11182
Date: 2020-02-10 07:43:56 +0000
From: @lersek

In order to keep bisectability, I&apos;d suggest something like:

- patch#1: introduce &quot;PcdRequireSelfSignedPk&quot; to SecurityPkg (code and PCD), with the PCD default being TRUE

- patch#2: explicitly set PcdRequireSelfSignedPk to TRUE in the OvmfPkg DSC files, when SECURE_BOOT_ENABLE

- patch#3: similarly in ArmVirtPkg

- patch#4: flip the DEC default in SecurityPkg to FALSE.

**Comment by tianocore-issues** [comment]
## Comment 11216
Date: 2020-02-11 19:16:18 +0000
From: Matthew Carlson &lt;<macarl>&gt;

I think we can go ahead with that plan. Perhaps #2 and #3 could be combined, but I can understand the case for them to be seperate.

The fix is public here: https://github.com/microsoft/mu_basecore/commit/0c424a7c0a505bbf81825b617a335cbbf455f091

**Comment by tianocore-issues** [comment]
## Comment 11230
Date: 2020-02-12 14:33:25 +0000
From: Matthew Carlson &lt;<macarl>&gt;

Created attachment 460
Potential patch for PK Self-signing spec

A potential patch that lines up with #1 on Lazlo&apos;s suggested patch list

Attachment: [BZ-460-pk_pcd_patch.txt](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-460-pk_pcd_patch.txt)

**Comment by tianocore-issues** [comment]
## Comment 11235
Date: 2020-02-12 20:16:12 +0000
From: nobody &lt;<nobody>&gt;

Matthew Carlson: can you share what test has been verified for this change?

**Comment by tianocore-issues** [comment]
## Comment 11236
Date: 2020-02-12 20:54:54 +0000
From: Matthew Carlson &lt;<macarl>&gt;

Hey Liming, 

I verified this works by clearing the PK and applying a new PK that was not self signed via the EFI shell and an OS call. I then verified that once set, the PK was not able to be set again, self-signed or not.

**Comment by tianocore-issues** [comment]
## Comment 13872
Date: 2020-10-02 19:00:46 +0000
From: @corthon

Any update on this?

**Comment by tianocore-issues** [comment]
## Comment 13881
Date: 2020-10-06 06:58:12 +0000
From: @lersek

(1) updating Liming&apos;s adddress on the CC list

(2) I think once Liming confirms that Matthew&apos;s testing described in comment 8 is okay, Matthew should post the series to edk2-devel.

**Comment by tianocore-issues** [comment]
## Comment 20534
Date: 2023-01-06 20:23:26 +0000
From: @jyao1

Would you please send out the patch set, we can continue the review process?

**Comment by tianocore-issues** [comment]
## Comment 20740
Date: 2023-01-20 18:08:17 +0000
From: Jan Bobek &lt;<jbobek>&gt;

I&apos;ve sent out a patch series to edk2-devel according to Laszlo&apos;s suggestion in comment #4, cover letter here:

https://edk2.groups.io/g/devel/message/98945

**Comment by tianocore-issues** [comment]
## Comment 20993
Date: 2023-02-04 06:54:22 +0000
From: @jyao1

Merged https://github.com/tianocore/edk2/pull/3998
